onnx/test/shape_inference_test.py

```diff
@@ -2992,6 +2992,22 @@ def test_negative_log_likehood_input_weight_shape_mismatch(self):  # type: () ->
             [make_node('NegativeLogLikelihoodLoss', ['input', 'target', 'weight'], ['loss'], reduction='none')],
             [])
         self.assertRaises(checker.ValidationError, self._inferred, graph)
+   
+   def test_cross_entropy_none(self):  # type: () -> None
+        graph = self._make_graph(
```

Comment:
We need some tests with N-D input.

Reply:
import torch
import torch.nn as nn
import numpy as np
from numpy.testing import assert_allclose

def softmaxcrossentropy(x, target, weight=None, reduction='mean'):  # type: (np.ndarray) -> np.ndarray
    max_x = np.max(x, axis=1, keepdims=True)
    exp_x = np.exp(x - max_x)
    p = exp_x / np.sum(exp_x, axis=1, keepdims=True)
    inp = np.log(p)
    input_shape = inp.shape
    if len(input_shape) == 2:
        N, C = input_shape
        neg_gather_element_input = np.zeros((N, ), dtype=np.float32)
        for i in range(N):
            neg_gather_element_input[i] = -inp[i][target[i]]
    if len(input_shape) == 3:
        N, C, D = input_shape
        neg_gather_element_input = np.zeros((N, D), dtype=np.float32)
        for i in range(N):
            for d in range(D):
                neg_gather_element_input[i][d] = -inp[i][target[i][d]][d]

    loss = neg_gather_element_input
    if weight is not None:
        gather_weight = np.take(weight, target)
        loss = gather_weight * loss
        if reduction == 'mean':
            return loss.sum() / gather_weight.sum()

    if reduction == 'mean':
        loss = np.mean(loss)
    if reduction == 'sum':
        loss = np.sum(loss)

    return loss

loss = nn.CrossEntropyLoss()
input = torch.randn(3, 5)
target = torch.empty(3, dtype=torch.long).random_(5)
output = loss(input, target)

my_loss = softmaxcrossentropy(input.numpy(), target.numpy())
assert_allclose(my_loss, output.numpy(), rtol=1e-05)

Reply:
Test numerical values:
```
import torch
import torch.nn as nn
import numpy as np
from numpy.testing import assert_allclose

def softmaxcrossentropy(x, target, weight=None, reduction='mean'):  # type: (np.ndarray) -> np.ndarray
    max_x = np.max(x, axis=1, keepdims=True)
    exp_x = np.exp(x - max_x)
    p = exp_x / np.sum(exp_x, axis=1, keepdims=True)
    inp = np.log(p)
    input_shape = inp.shape
    if len(input_shape) == 2:
        N, C = input_shape
        neg_gather_element_input = np.zeros((N, ), dtype=np.float32)
        for i in range(N):
            neg_gather_element_input[i] = -inp[i][target[i]]
    if len(input_shape) == 3:
        N, C, D = input_shape
        neg_gather_element_input = np.zeros((N, D), dtype=np.float32)
        for i in range(N):
            for d in range(D):
                neg_gather_element_input[i][d] = -inp[i][target[i][d]][d]

    loss = neg_gather_element_input
    if weight is not None:
        gather_weight = np.take(weight, target)
        loss = gather_weight * loss
        if reduction == 'mean':
            return loss.sum() / gather_weight.sum()

    if reduction == 'mean':
        loss = np.mean(loss)
    if reduction == 'sum':
        loss = np.sum(loss)

    return loss

loss = nn.CrossEntropyLoss()
input = torch.randn(3, 5)
target = torch.empty(3, dtype=torch.long).random_(5)
output = loss(input, target)

my_loss = softmaxcrossentropy(input.numpy(), target.numpy())
assert_allclose(my_loss, output.numpy(), rtol=1e-05)
```